### ISSUE TITLE
Ruff complains wrongly when we delete a variable in a function ( F821 )

### ISSUE THREAD

**Issue by Leo-Adlakha** [astral-sh/ruff#9858]
Hi, Ruff raises an issue with the following code snippet, even when the code is running fine. Could you fix this issue ?

Code Snippet for reproducing the bug:

```
def abc():
    global ABC_123
    global XYZ_123

    class ABC_123():
        pass

    class XYZ_123():
        def inputs(self):
            yield ABC_123

    print(XYZ_123().inputs())

    del ABC_123
    del XYZ_123

abc()
```

Command invoked:

```
ruff <name-of-the-file>
```

Ruff version - `0.1.6`

**Comment by charliermarsh** [comment]
This likely won't be fixed in the near future since it requires tracing the function calls themselves -- i.e., we need to know that `inputs()` wasn't called after the variable was deleted, as in:

```python
def abc():
    global ABC_123
    global XYZ_123

    class ABC_123():
        pass

    class XYZ_123():
        def inputs(self):
            yield ABC_123

    instance = XYZ_123()

    del ABC_123

    instance = XYZ_123()

abc()
```

(Mypy and Pyright don't raise any errors on this, even though it will error at runtime.)

I'll label it as a bug but candidly I'd recommend trying to restructure the code to avoid this kind of binding flow -- it's hard to follow and really hard for static analyzers to help you with.

**Comment by Avasam** [comment]
Is this a duplicate of either/or:
- https://github.com/astral-sh/ruff/issues/6242
- https://github.com/astral-sh/ruff/issues/7733
- https://github.com/astral-sh/ruff/issues/9349

And the similar tracing problem as with other `F821` issues?
